### ISSUE TITLE
Connection aborted with streams and request guards

### ISSUE THREAD

**Issue by super-cilious** [SergioBenitez/Rocket#892]
hey, I'm using Rocket 0.4.0, Windows 10 (1809), and getting ConnectionAborted (An established connection was aborted by the software in your host machine.) when reading ``Data`` alongside request guards; here's the log:

```
POST /api/user/avatar multipart/form-data; boundary=----WebKitFormBoundaryB9kh66R5Vm4cOKBE:
    => Matched: POST /api/<path..> (post_api)
    => Outcome: Forward
    => Matched: POST /api/user/avatar (set_avatar_route)
    => Warning: Data left unread. Force closing network stream.
    => Outcome: Failure
    => Warning: Responding with 429 Too Many Requests catcher.
    => Error: Failed to write response: Os { code: 10053, kind: ConnectionAborted, message: "An established connection was aborted by the software in your host machine." }.
```

the "data left unread" is probably relevant, but i see that it should only ``Shutdown::Read`` (something similar that was resolved in #386 ), so i don't know why rocket has trouble writing a response

here is the function header:
```
#[post("/avatar", data="<data>")]
fn set_avatar_route(_up: Catch<UploadFile>, auth: VerAuth, conn: RocketConn, multipart: Multipart, data: Data) -> ApiRes<()> {
```
there are a bunch of request guards that should be able to route to catchers which can write responses

**Comment by jebrosen** [comment]
Is there any chance you could try the same code on another platform, or provide a full but less involved example that also fails that someone else could test? It would be good to know if it fails only on Windows. Otherwise, I can try to make a reproduction when I get a chance.

**Comment by Thomas-QM** [comment]
thanks for the hint! yeah it works fine on ubuntu...

**Comment by jebrosen** [comment]
Is this still a priority? I never got around to testing this on windows, but I probably could soon.

**Comment by Thomas-QM** [comment]
if you wanna bang out a bug go ahead, but its not a priority over here

**Comment by jebrosen** [comment]
For what it's worth, I have been able to reproduce this error in 0.4 by using the code in #386:

```
> POST / HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.61.1
> Accept: */*
> Content-Length: 274063360
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
} [65536 bytes data]
* Recv failure: Connection was reset
  0  261M    0     0    0  576k      0   562M --:--:-- --:--:-- --:--:--  562M
* Closing connection 0
curl: (56) Recv failure: Connection was reset
```

But the in-progress port of Rocket to hyper 0.12 works as desired:

```
> POST / HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.61.1
> Accept: */*
> Content-Length: 274063360
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
} [65536 bytes data]
< HTTP/1.1 401 Unauthorized
< content-type: text/html; charset=utf-8
< server: Rocket
< content-length: 501
< date: Sat, 01 Jun 2019 05:52:43 GMT
* HTTP error before end of send, stop sending
<
{ [501 bytes data]
  1  261M  100   501    1 2752k  31312   167M  0:00:01 --:--:--  0:00:01  167M
            <!DOCTYPE html>
            <html>
            <head>
                <meta charset="utf-8">
                <title>401 Unauthorized</title>
            </head>
            <body align="center">
                <div align="center">
                    <h1>401: Unauthorized</h1>
                    <p>The request requires user authentication.</p>
                    <hr />
                    <small>Rocket</small>
                </div>
            </body>
            </html>

* Closing connection 0
```

So, #1008 and wherever that work continues will probably just fix this issue along the way.

**Comment by SergioBenitez** [comment]
@jebrosen Does `master` resolve the issue?

**Comment by jebrosen** [comment]
Yes - still using the code in #386 (updated to the appropriate APIs), with both tests done on Windows:

* v0.4.5 (4683407): Server closes the connection and later logs `Error: Failed to write response: Os { code: 10053, kind: ConnectionAborted, message: "An established connection was aborted by the software in your host machine." }.` Client logs `curl: (56) Recv failure: Connection was reset`. 
* master (45b4436): Server does not close the connection, and successfully sends the 401. Client logs `* HTTP error before end of send, stop sending`.

hyper (since 0.13.4) includes a fix that looks equivalent to what rocket 0.4 did to attempt to drain the body: hyperium/hyper#2149. One key difference from 0.4 is that it now happens *after* the Response is sent to the client. Any further issues in this area will go through different code paths / need different fixes and should be opened as new issues.
